extensions/amp-story-panning-media/0.1/amp-story-panning-media.js

```diff
@@ -167,17 +191,68 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
   }
 
   /** @private */
-  onPageNavigation_() {
-    if (this.isOnActivePage_) {
-      this.animate_();
+  setAnimateTo_() {
+    const x = parseFloat(this.element_.getAttribute('x') || 0);
+    const y = parseFloat(this.element_.getAttribute('y') || 0);
+    const zoom = parseFloat(this.element_.getAttribute('zoom') || 1);
+    const lockBounds = this.element_.hasAttribute('lock-bounds');
+
+    if (lockBounds) {
+      // Zoom must be set to calculate maxBounds.
+      this.animateTo_.zoom = zoom < 1 ? 1 : zoom;
+      const maxBounds = this.getMaxBounds_();
+      this.animateTo_.x =
+        x > 0 ? Math.min(maxBounds.left, x) : Math.max(maxBounds.right, x);
+      this.animateTo_.y =
+        y > 0 ? Math.min(maxBounds.top, y) : Math.max(-maxBounds.bottom, y);
+    } else {
+      this.animateTo_ = {x, y, zoom};
+    }
+  }
+
+  /**
+   * Calculates max distances to keep image in viewport.
+   * @private
+   * @return {panningMediaMaxBoundsPercentDef}
+   */
+  getMaxBounds_() {
+    // Calculations to clamp image to edge of container.
+    const {width: containerWidth, height: containerHeight} = this.pageSize_;
+
+    const ampImgWidth = this.ampImgEl_.getAttribute('width');
+    const ampImgHeight = this.ampImgEl_.getAttribute('height');
+    if (!ampImgWidth || !ampImgHeight) {
+      user().error(
+        TAG,
+        '"lock-bounds" requires "width" and "height" to be set on the amp-img child.'
+      );
     }
+    // TODO(#31515): When aspect ratio is portrait, containerWidth will be used for this.
+    const percentScaled = containerHeight / ampImgHeight;
```

Comment:
Could we use the `zoom` variable here? I think it would have the same value unless I'm missing sth (percentScaled = zoom right?)

Reply:
It's the amount the image was scaled to fit the viewport. I renamed the variable.

There might be a better way to calculate this. I have a feeling comparing the viewport aspect ratio and image aspect ratio could be another approach but I'm not sure.